rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,105 @@
+---
+feature: Defined conventions around testing of official Nixpkgs packages.
+start-date: 2021-12-29
+author: Jonathan Ringer
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+  - [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88)
+---
+
+# Summary
+[summary]: #summary
+
+When updating or modifying packages, several conventions for testing regressions
+have been adopted. However, these practices are not standard, and generally it's not well
+defined how each testing method should be implemented. It would be beneficial to have
+an unambiguous way to say that a given package, and all downstream dependencies, have
+had as many automated test ran possible. This will give a high degree of certainty that
+a given change is less likely to manifest regressions once introduced on a release
+channel.
+
+Another desire of this rfc is also to have a way for various review tools
+(e.g. ofborg, hydra, nixpkgs-review) to have a standard way to determine if a
+package has additional tests which can help verify its correctness.
+
+# Motivation
+[motivation]: #motivation
+
+Breakages are a constant painpoint for nixpkgs. It is a very poor user experience to
+have a configuration broken because one or more packages fail to build. Often when
+these breakages occur, it is because the change had a large impact on the entirety
+of nixpkgs; and unless there's a dedicated hydra jobset for the pull request, it's
+infeasible to expect pull request authors to verify every package affected
+by a change they are proposing. However, it is feasible to specify packages that
+are very likely to be affected by changes in another package, and use this information
+to help mitigate regressions from appearing in release channels.
+
+# Detailed design
+[design]: #detailed-design
+
```

Comment:
Think this technique was @jonringer 's idea in the first place :wink:

The main thing I'm slightly uncomfortable with is adding the reverse dependencies as arguments - I imagine it causing wierdness for people who do a lot of overriding. But I can't think of a sensible way it could be avoided.

Reply:
Ideally the benefit is for nixpkgs testing. For the overriding case, it should be that an override would not trigger a downstream dependency to influence the build as passthru gets pruned before instantiation.

I agree it's odd that a package is aware of it's downstream dependencies, but I'm not sure of another way of having a package be aware of them.

Reply:
usage of `pkgs` is frowned upon in nixpkgs from a hygiene perspective, I would rather not involve it

Reply:
**nixpkgs-review** builds all reverse dependencies. could we just use this mechanism in a `testers.testBuildReverseDeps`? but there can be thousands :grimacing: 

cc @Mic92